src/app/clusters/commodity-metering-server/commodity-metering-server.cpp

```diff
@@ -0,0 +1,360 @@
+                continue;
+            }
+            return false;
+        }
+        return true;
+    }
+    return false;
+}
+
+template <typename T>
+struct SpanCopier
+{
+    static bool Copy(const chip::Span<const T> & source, DataModel::List<const T> & destination,
+                     size_t maxElements = std::numeric_limits<size_t>::max())
+    {
+        if (source.empty())
+        {
+            destination = DataModel::List<const T>();
+            return true;
+        }
+
+        size_t elementsToCopy = std::min(source.size(), maxElements);
+        auto * buffer         = static_cast<T *>(chip::Platform::MemoryCalloc(elementsToCopy, sizeof(T)));
+
+        if (!buffer)
+        {
+            return false;
+        }
+
+        std::copy(source.begin(), source.begin() + elementsToCopy, buffer);
+        destination = DataModel::List<const T>(chip::Span<const T>(buffer, elementsToCopy));
+        return true;
+    }
+};
+} // namespace
+
+static void CleanUpIDs(DataModel::List<const uint32_t> & IDs)
+{
+    if (!IDs.empty() && IDs.data())
+    {
+        Platform::MemoryFree(const_cast<uint32_t *>(IDs.data()));
+        IDs = DataModel::List<const uint32_t>();
+    }
+}
+
+static void CleanupMeteredQuantityData(DataModel::List<Structs::MeteredQuantityStruct::Type> & aValue)
+{
+    if (aValue.data() != nullptr)
+    {
+        for (auto & item : aValue)
+        {
+            CleanUpIDs(item.tariffComponentIDs);
+        }
+
+        Platform::MemoryFree(aValue.data());
+        aValue = DataModel::List<Structs::MeteredQuantityStruct::Type>();
+    }
+}
+
+Instance::~Instance()
+{
+    if (!mMeteredQuantity.IsNull())
+    {
+        CleanupMeteredQuantityData(mMeteredQuantity.Value());
+    }
+    Shutdown();
+}
+
+CHIP_ERROR Instance::Init()
+{
+    mMeteredQuantity.SetNull();
+    mMeteredQuantityTimestamp.SetNull();
+    mMeasurementType.SetNull();
+    mMaximumMeteredQuantities.SetNull();
+    VerifyOrReturnError(AttributeAccessInterfaceRegistry::Instance().Register(this), CHIP_ERROR_INCORRECT_STATE);
+    return CHIP_NO_ERROR;
+}
+
+void Instance::Shutdown()
+{
+    AttributeAccessInterfaceRegistry::Instance().Unregister(this);
+}
+
+static CHIP_ERROR CopyMeteredQuantityEntry(const Structs::MeteredQuantityStruct::Type & src,
+                                           Structs::MeteredQuantityStruct::Type & dest)
+{
+    dest.quantity = src.quantity;
+
+    if (!SpanCopier<uint32_t>::Copy(src.tariffComponentIDs, dest.tariffComponentIDs, kMaxTariffComponentIDsPerMeteredQuantityEntry))
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::SetMeteredQuantity(const DataModel::Nullable<DataModel::List<Structs::MeteredQuantityStruct::Type>> & newValue)
+{
+    assertChipStackLockedByCurrentThread();
+
+    if (NullableListEqual(newValue, mMeteredQuantity))
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    if (newValue.IsNull())
+    {
+        mMeteredQuantity.SetNull();
```

Comment:
This is a memory leak, no, in the case when the old value is not null and the new value is?  I would think this block:
```
        if (!mMeteredQuantity.IsNull())
        {
            CleanupMeteredQuantityData(mMeteredQuantity.Value());
        }
```

should just be right after the block guarded by the NullableListEqual check.